### PR TITLE
changing fetch to return an array of css strings

### PR DESCRIPTION
**Pull request by oluoluoxenfree**
## Steps to test/reproduce
_Please explain how to best reproduce the issue and/or test the changes locally (including the pages/URLs/views/states to review)._

## Show me
_Provide screenshots/animated gifs/videos if necessary._

# REMEMBER: Attach this PR to the Trello card

### CODE REVIEW

File: src/fetching.ts [hunk 1]
```diff
@@ -1,21 +1,32 @@
...
       CSSlinks.push(link);
     }
   });
-  return CSSlinks;
+
+  const linkedCss = CSSlinks.map((link) => {
+    // fetch css and push into array of strings
+    console.log(link.href);
+    http.get(link.href).toString();
+  });
```
Comment (jgerigmeyer) [inline src/fetching.ts:19]: This `.map` call needs to return something for each item, otherwise it results in e.g. `[undefined]`.

File: src/fetching.ts [hunk 2]
```diff
@@ -1,21 +1,32 @@
...
+    if ((link.type === 'text/css' || link.rel === 'stylesheet') && link.href) {
       CSSlinks.push(link);
     }
   });
-  return CSSlinks;
+
+  const linkedCss = CSSlinks.map((link) => {
+    // fetch css and push into array of strings
+    console.log(link.href);
+    http.get(link.href).toString();
```
Comment (jgerigmeyer) [inline src/fetching.ts:18]: `http.get()` is asynchronous, so to get the string here we'll need to wait for response data, right?

https://nodejs.org/docs/latest-v16.x/api/http.html#httpgeturl-options-callback

File: index.html [hunk 3]
```diff
@@ -21,7 +21,7 @@ <h1>CSS Anchor Positioning Polyfill</h1>
     </popup>
 
     <script type="module">
-      import transformCSS from '/src/index.ts';
+      import transformCSS from './dist/css-anchor-positioning.js';
```
Comment (jgerigmeyer) [inline index.html:24]: Why is this change necessary?

File: index.html [hunk 4]
```diff
@@ -5,8 +5,8 @@
     <meta http-equiv="X-UA-Compatible" content="IE=edge" />
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
     <title>CSS Anchor Positioning Polyfill Demo</title>
-    <link rel="stylesheet" href="/src/base.css" />
-    <link type="text/css" href="/src/styles.css" />
+    <link rel="stylesheet" href="./src/base.css" />
+    <link type="text/css" href="./src/styles.css" />
```
Comment (jgerigmeyer) [inline index.html:9]: Were the absolute URLs not working here?

File: src/fetching.ts [hunk 5]
```diff
@@ -2,31 +2,52 @@ import { computePosition } from '@floating-ui/dom';
...
-  return CSSlinks;
+
+  const linkedCSS = await Promise.all(
+    CSSlinks.map(async (link) => {
+      // fetch css and push into array of strings
+      const response = await fetch(link.href);
+      const text = await response.text();
+      return text;
+    }),
+  );
```
Comment (jgerigmeyer) [inline src/fetching.ts:21]: @oluoluoxenfree When using `Array.prototype.map()` with an asynchronous callback, the array is an array of Promises -- so in this case we want to wrap with `await Promise.all()` to get the underlying resolved values.

https://stackoverflow.com/a/40140562

File: src/index.ts [hunk 6]
```diff
@@ -1,3 +1,6 @@
+// Polyfill `fetch` for older browsers
+import 'whatwg-fetch';
```
Comment (jgerigmeyer) [inline src/index.ts:2]: I'm not actually sure if this is needed or not? `fetch` is supported by all ES6 browsers AFAIK, and that's our current support target.

File: index.html [hunk 7]
```diff
@@ -5,8 +5,13 @@
...
     <title>CSS Anchor Positioning Polyfill Demo</title>
-    <link rel="stylesheet" href="/src/base.css" />
-    <link type="text/css" href="/src/styles.css" />
+    <link rel="stylesheet" href="/base.css" />
+    <link type="text/css" href="/styles.css" />
+    <style>
+      .test {
+        color: red;
+      }
+    </style>
```
Comment (jgerigmeyer) [inline index.html:14]: @oluoluoxenfree Feel free to remove/change this -- I just wanted to test that this was functionally working.

File: index.html [hunk 8]
```diff
@@ -5,8 +5,13 @@
     <meta http-equiv="X-UA-Compatible" content="IE=edge" />
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
     <title>CSS Anchor Positioning Polyfill Demo</title>
-    <link rel="stylesheet" href="/src/base.css" />
-    <link type="text/css" href="/src/styles.css" />
+    <link rel="stylesheet" href="/base.css" />
+    <link type="text/css" href="/styles.css" />
```
Comment (jgerigmeyer) [inline index.html:9]: @oluoluoxenfree The main fix was moving these styles into the `public/` dir. By default, Vite does a lot of fancy pre-processing and auto-injection when CSS files are imported or fetched. By moving them to `public/` we can tell Vite to serve these assets raw, which is I think what we want for this polyfill.

See:
- https://vitejs.dev/guide/features.html#css
- https://vitejs.dev/guide/assets.html#the-public-directory

File: src/fetching.ts (hunk 5 above)
Comment (jgerigmeyer) [reply src/fetching.ts:21]: It strikes me that most of these CSS polyfills would be doing something similar -- finding all the styles in the document, and extracting them as strings. Is there code in the other existing polyfills we could copy?

File: package.json [hunk 9]
```diff
@@ -91,7 +93,8 @@
   },
   "yarn-upgrade-all": {
     "ignore": [
-      "@types/node"
+      "@types/node",
+      "node-fetch"
```
Comment (sanajaved7) [inline package.json:97]: I think you said we didn't need node-fetch after all right?

File: src/fetching.ts [hunk 10]
```diff
@@ -2,10 +2,14 @@ import { computePosition } from '@floating-ui/dom';
 
 import type { PositionFallbackRulesMap } from './parsing.js';
 
+export const CSSlinks: HTMLLinkElement[] = [];
```
Comment (jgerigmeyer) [inline src/fetching.ts:5]: @oluoluoxenfree I'm curious why this array is moved top-level and exported?

File: package.json (hunk 9 above)
Comment (jgerigmeyer) [reply package.json:97]: This is required to run the tests -- what we didn't need (and Olu already removed) was the production whatwg-fetch polyfill.

File: src/fetching.ts (hunk 10 above)
Comment (oluoluoxenfree) [reply src/fetching.ts]: @sanajaved was saying preserving the order and destination of the links was important to what she was doing, so I was making sure it's available. is there an easier way to do that?

**Review by jgerigmeyer — APPROVED** [review]

File: src/fetching.ts (hunk 10 above)
Comment (sanajaved7) [reply src/fetching.ts]: +1 to Olu and for some more context - when transforming the original styles in a stylesheet - it would helpful to know the href of the stylesheet that particular CSS string came from since after transformation, I need to replace that href URL with the new source URL. I could probably rework this for what I need though if we want to go ahead and merge this as-is. But I was hoping we wouldn't need to write two separate CSS fetching functions.

Comment (jgerigmeyer) [reply src/fetching.ts]: Ah, ok. What about including the src href in the returned list? Something like https://github.com/oddbird/css-anchor-positioning/pull/13/commits/9dd04517ef991db1a07f43b5bd4afe7b23117729

Comment (sanajaved7) [reply src/fetching.ts]: That's perfect - thanks! 🎉

**Comment by sanajaved7** [comment]
[Make fetchCSS command return (an array/object of) strings](https://trello.com/c/gbU3qRHx/43-make-fetchcss-command-return-an-array-object-of-strings)

**Review by sanajaved7 — APPROVED** [review]

**Review by jgerigmeyer — APPROVED** [review]
